packets2/connack.go

```diff
@@ -0,0 +1,84 @@
+package packets2
+
+import (
+	"encoding/binary"
+	"fmt"
+
+	pkts "github.com/energomonitor/bisquitt/packets"
+)
+
+const connackHeaderLength uint16 = 6
+const connackSessionPresentBit uint8 = 1
+
+type Connack struct {
+	pkts.Header
+	PacketV2
+	// Flags
+	SessionPresent bool
+	// Fields
+	SessionExpiryInterval    uint32
+	ReasonCode               ReasonCode
+	AssignedClientIdentifier string
+}
+
+func NewConnack(sessExpiry uint32, reasonCode ReasonCode, assignedClientID string,
+	sessPresent bool) *Connack {
+	p := &Connack{
+		Header:                   *pkts.NewHeader(pkts.CONNACK, 0),
+		SessionPresent:           sessPresent,
+		SessionExpiryInterval:    sessExpiry,
+		ReasonCode:               reasonCode,
+		AssignedClientIdentifier: assignedClientID,
+	}
+	p.computeLength()
+	return p
+}
+
+func (p *Connack) computeLength() {
+	assignedClientIDLength := uint16(len(p.AssignedClientIdentifier))
+	p.Header.SetVarPartLength(connackHeaderLength + assignedClientIDLength)
+}
+
+func (p *Connack) decodeFlags(b byte) {
+	p.SessionPresent = (b & connackSessionPresentBit) == connackSessionPresentBit
+}
+
+func (p *Connack) encodeFlags() byte {
+	var b byte
+	if p.SessionPresent {
+		b |= connackSessionPresentBit
+	}
+	return b
+}
+
+func (p *Connack) Pack() ([]byte, error) {
+	buf := p.Header.PackToBuffer()
+
+	_ = buf.WriteByte(byte(p.ReasonCode))
+	_ = buf.WriteByte(p.encodeFlags())
+	_, _ = buf.Write(pkts.EncodeUint32(p.SessionExpiryInterval))
+	if p.AssignedClientIdentifier != "" {
+		_, _ = buf.Write([]byte(p.AssignedClientIdentifier))
+	}
+
+	return buf.Bytes(), nil
+}
+
+func (p *Connack) Unpack(buf []byte) error {
+	if len(buf) < int(connackHeaderLength) {
+		return fmt.Errorf("bad CONNACK2 packet length: expected >=%d, got %d",
+			connackHeaderLength, len(buf))
+	}
+
+	p.ReasonCode = ReasonCode(buf[0])
+	p.decodeFlags(buf[1])
+	p.SessionExpiryInterval = binary.BigEndian.Uint32(buf[2:6])
+	p.AssignedClientIdentifier = string(buf[6:])
```

Comment:
In the spec, the Session Expiry Interval and Assigned Client Identifier fields are marked as optional. It should be reflected here.

Reply:
See previous comment. I consider `Session Expiry Interval` mandatory, therefore it's always unpacked. If `Assigned Client Identifier` is not present, it will just unpack as a zero-length string, which is the expected behavior.

Reply:
Makes sense. Please add a comment here too.